[EN] update GitOps (#1334)
cncf/glossary: content/en/gitops.md

---
title: GitOps
status: Completed
category: Concept
tags: ["methodology"]
---

## What it is

> GitOps is a set of best practices based on [shared principles](https://github.com/open-gitops/documents/blob/release-v1.0.0/PRINCIPLES.md), applied to a workflow that depends on software agents to continuously and automatically reconcile an actual running system with its desired state declared in an immutably versioned state store like Git.

Reviewer: I'm afraid this is really hard to understand for non-technical people. What are workflows and agents? The concept of reconciling a system is also new. Combined with the other two terms makes it really challenging for people not familiar with these concepts to understand what we are talking about. That's why I tried to simplify it in the Google doc to something more like this. 

```suggestion
GitOps is a set of best practices based on [shared principles](https://github.com/open-gitops/documents/blob/release-v1.0.0/PRINCIPLES.md) that ensure the application and system infrastructure always reflect the desired state
```
Isn't that the essence of what that means but without all these industry terms? Let's keep it really simple and avoid introducing other technical terms if we don't have to. 

This was new in the Google doc, but it feels like it really helps understand what we are talking about. Can we add it?

> While [DevOps](https://glossary.cncf.io/devops/) focuses on a cultural shift, GitOps is a concrete set of deployment practices that allow adopters to realize the DevOps promise.

Reviewer 2: Still, we need to add that the desired state is reflected in a git repository (hence "git"ops), so perhaps

```suggestion
GitOps is a set of best practices based on [shared principles](https://github.com/open-gitops/documents/blob/release-v1.0.0/PRINCIPLES.md) that ensure the application and system infrastructure always reflect their desired state declared beforehand in a single point of truth (usually a Git repository).
```

> These software agents and practices are used to execute a cohesive workflow and extends this practice to applications, infrastructure, and operational procedures.

Reviewer: What does a "cohesive workflow" and "extending it to..." mean? Also, maybe this should go in the "how" section?

## Problem it addresses

> Existing processes for infrastructure configuration management – including scripting and tightly coupled CI/CD – are often error prone and can face challenges such as configuration drift, failed deployments, relying on a system's previous state for success, missing documentation, or unknown development history.

Reviewer: ```suggestion
Existing processes for infrastructure configuration management — including scripting and tightly coupled [CI](https://glossary.cncf.io/continuous-integration/)/[CD](https://glossary.cncf.io/continuous-delivery/) — are often error-prone and face many challenges. 
They often lead to configuration drift, failed deployments, relying on a system's previous state for success, missing documentation, or unknown development history.
```

Broke sentence down for better readability. We have an entry for continuous deployment and delivery. Is linking to one better than to another?

Reviewer 2: Please not introduce 'configuration drift ' as a new term. 

```suggestion
Existing processes for infrastructure configuration management — including scripting and tightly coupled [CI](https://glossary.cncf.io/continuous-integration/)/[CD](https://glossary.cncf.io/continuous-delivery/) — are often error-prone and face many challenges. 

They often lead to failed deployments, relying on a system's previous state for success, missing documentation, or unknown development history.
```

> Event driven CD does not address drift that may happen independently of CI.

Reviewer: What does this mean?

> Giving external automation tools credentials to your production systems widens the attack footprint.

Reviewer: When and why would you give credentials your automation tools? Seems like context is missing here.

Adopting a GitOps workflow can help alleviate these issues, among several others.

## How it helps

> The GitOps paradigm can be applied to a workflow to help manage applications and cloud system infrastructure. 

Reviewer 2: This doesn't answer the question 'How it helps' - it only says it can be applied.

Reviewer: Can we say methodology instead of paradigm? That way, we keep it simple. Paradigm is a little more abstract and confusing. 

What does applying workflows to apps and infrastructure mean? We probably need to define workflow above (or use another term) to use it here. And to @wojtekidd's point, we'd need to add: "...help manage apps and cloud system infrastructure by...[the how it helps is missing]"

> It enables organizations several advantages such as better coordination, transparency, stability, and reliability of a system, and faster mean time to recovery.

Reviewer 3: Advantages im comparison to what? I think there are many things that create the mentioned advantages. "Simple" CICD does that as well. IMHO the specific advantage of GitOps in comparison to "simple" CICD should be in focus.

> Direct access to desired state store and continuous reconciliation allows GitOps agents to ensure the current live state of a system matches the desired target state, regardless of when or how any system failure or drift between the two may occur.

Reviewer: Some of what is here, is captured in the Google doc but in simpler language. Can repurpose some of this?

> Traditionally seen as a single process, GitOps decouples CI/CD (principles 3 and 4). While [CI](https://glossary.cncf.io/continuous-integration/) is focused on building applications iteratively, CD focuses on seeing that change in the target system. GitOps concentrates on the CD aspect — both [delivery](https://glossary.cncf.io/continuous-delivery/) and [deployment](https://glossary.cncf.io/continuous-deployment/). That allows developers to focus on CI without worrying about how the change ends up on the target system — they just need to know that it does.
> Configuration files, stored as code in a repository ([infrastructure as code](https://glossary.cncf.io/infrastructure-as-code/)), generate the same infrastructure environment each time it is deployed, creating consistency. And to ensure what's running on the cluster doesn't diverge from the desired state (codified in Git), [software agents](https://www.britannica.com/technology/software-agent) continuously monitor the live environment. The agent will attempt to roll back or match the desired state if divergence is detected, avoiding configuration drift.

> Using GitOps for CD improves security by separation of duty between CI and CD, where CI no longer needs credentials to your live system.

Reviewer 2: I think you meant CD tools?

```suggestion
Using GitOps for CD improves security by separation of duty between CI and CD, where CD tooling no longer needs credentials to your live system.
```
